### PR TITLE
[llvm][AsmPrinter] Add an option to print instruction latencies

### PR DESCRIPTION
**Pull request by jroelofs**
... matching what we have in the disassembler.  This isn't turned on by default since several of the scheduling models are not completely accurate, and we don't want to be misleading.

### CODE REVIEW

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 1]
```diff
@@ -1080,8 +1087,78 @@ void AsmPrinter::emitFunctionEntryLabel() {
...
+    // have a scheduling table.  Note the default does not have a table.
+    return getItineraryLatency(MI, MF, STI);
+
+  // Get the scheduling class of the requested instruction.
+  const TargetInstrInfo *TII = MF->getSubtarget().getInstrInfo();
+  const MCInstrDesc &Desc = TII->get(MI.getOpcode());
+  unsigned SCClass = Desc.getSchedClass();
+  const MCSchedClassDesc *SCDesc = SCModel.getSchedClassDesc(SCClass);
+  // Resolving the variant SchedClass requires an MI to pass to
+  // SubTargetInfo::resolveSchedClass.
```
Comment (davemgreen) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1142]: We could have an MI in this case?

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 2]
```diff
@@ -1080,8 +1087,78 @@ void AsmPrinter::emitFunctionEntryLabel() {
...
+                               const MCSubtargetInfo *STI) {
+  const int NoInformationAvailable = -1;
+  const TargetInstrInfo *TII = MF->getSubtarget().getInstrInfo();
+
+  // Check if we have a CPU to get the itinerary information.
+  if (STI->getCPU().empty())
+    return NoInformationAvailable;
+
+  // Get itinerary information.
+  InstrItineraryData IID = STI->getInstrItineraryForCPU(STI->getCPU());
```
Comment (davemgreen) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1105]: Is it worth adding a test for a CPU that has an Itinerary model, but no InstrSchedModel?

**Review by davemgreen — COMMENTED** [review]
Would it be possible to share the code with the existing implementation in Disassembler.cpp?

**Review by qcolombet — COMMENTED** [review]
> https://github.com/llvm/llvm-project/pull/113243#pullrequestreview-2384034868

+1 on this

E.g., templatize the getLatency function on MachineInstr/MCInst and provide a functor to call `SubTargetInfo::resolveSchedClass`.

**Comment by jroelofs** [comment]
> > [#113243 (review)](https://github.com/llvm/llvm-project/pull/113243#pullrequestreview-2384034868)
> 
> +1 on this
> 
> E.g., templatize the getLatency function on MachineInstr/MCInst and provide a functor to call `SubTargetInfo::resolveSchedClass`.

I am not sure where to put the common thing after adding overloads for that: `MCSchedModel` seems like the obvious place, but putting it there would be a layering violation, since it would cause `MC` to depend on `CodeGen`.

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp (hunk 2 above)
Comment (jroelofs) [reply llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:1105]: done

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp (hunk 1 above)
Comment (jroelofs) [reply llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp]: copy-pasta from the one in the disassembler, where it only had a MCInst and not an MI.

**Comment by qcolombet** [comment]
> > > [#113243 (review)](https://github.com/llvm/llvm-project/pull/113243#pullrequestreview-2384034868)
> > 
> > 
> > +1 on this
> > E.g., templatize the getLatency function on MachineInstr/MCInst and provide a functor to call `SubTargetInfo::resolveSchedClass`.
> 
> I am not sure where to put the common thing after adding overloads for that: `MCSchedModel` seems like the obvious place, but putting it there would be a layering violation, since it would cause `MC` to depend on `CodeGen`.

The instantiation of the template can be in the right library, no?
And the functor I was mentioning would be nullptr by default and only implemented in the CodeGen library.

**Comment by jroelofs** [comment]
> The instantiation of the template can be in the right library, no?
> And the functor I was mentioning would be nullptr by default and only implemented in the CodeGen library.

Ok, that works.

**Review by davemgreen — APPROVED** [review]
From what I can tell this LGTM. Cheers

**Pull request merged**